src/abstractitensornetwork.jl

```diff
@@ -721,6 +721,48 @@ function contract_inner(
   return contract(tn; sequence)[]
 end
 
+function contract_with_BP(
+  itn::AbstractITensorNetwork; outputlevel=1, partitioning=group(v -> v, vertices(itn))
+)
+  return contract_with_BP(
+    ComplexF64, itn::AbstractITensorNetwork; outputlevel, partitioning
+  )
+end
+
+function contract_with_BP(
+  T::Type,
```

Comment:
I'm confused by this part of the interface, why can you set the element type in this way?

Reply:
This is not the right way to do it. Basically, due to the `log` in this implementation, we need to promote the scalars of which the element type is taken to `complex`. Then the idea was that `contract_with_BP` returns a complex scalar, unless you specify a type, in which case it converts the scalar to the specified type. This should of course be handled differently for a proper implementation as `contract` with `BP` backend.

Reply:
Ok. I think `contract_with_BP` as it stands is too far away from being a library function, so we should think about how to deal with that before continuing forward with this PR.